### PR TITLE
Media change listener

### PR DESCRIPTION
**Pull request by Inwerpsel**
This makes the component set the embed html in the input change listener so that we don't need to set the attributes in an intermediary invalid state until the effect runs, but instead sets both at the same time. This ensures that no render can unintentionally trigger a change to the post content.

Since the backend will still render any block that has no content**, we can make the block return `null` in save whenever it's missing either (or both) `embed_html` or `poster_url`, but does have the corresponding "identifier" prop (`media_url` or `video_poster_img`). Then it uses [`useSelect` in the edit component](https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/423/files#diff-c6c91408622162ae2bd090b6d50b303d1de0a3ebff4e385997189643c4a18ce1R124) to get the missing html or url.

We can in a follow up make a migration script to migrate existing content, then after running it we only need to remove a few lines from `MediaEditor` and `save`.

_** well technically it adds the missing attributes for the frontend to render_

### CODE REVIEW

File: assets/src/blocks/Media/MediaEditor.js [hunk 1]
```diff
@@ -17,10 +17,24 @@ const MediaInspectorOptions = ({ attributes, setAttributes }) => {
   // https://github.com/facebook/react/issues/14904
   // https://github.com/facebook/react/issues/955#issuecomment-469352730
   const [ media_url, setMediaURL ] = useState(attributes.media_url);
-  const debouncedMediaURLUpdate = useCallback(debounce(value => setAttributes({ media_url: value }), 300), []);
+  const debouncedMediaURLUpdate = useCallback(debounce(value => {
+    setAttributes({ embed_html: null });
```
Comment (pablocubico) [inline assets/src/blocks/Media/MediaEditor.js:21]: Why the need to set `null` here?

Comment (Inwerpsel) [reply assets/src/blocks/Media/MediaEditor.js:21]: We probably don't, I added it so that the previous embed html is already cleared before the api request to get the embed is completed, but I guess we can just leave the previous html in place until the new arrives.

File: assets/src/blocks/Media/MediaEditor.js [hunk 2]
```diff
@@ -17,10 +17,24 @@ const MediaInspectorOptions = ({ attributes, setAttributes }) => {
...
 
-  function onSelectImage({id}) {
-    setAttributes({video_poster_img: id});
+    const embedPreview = wp.data.select('core').getEmbedPreview(resolveURL(value));
+
+    const embed_html = embedPreview ? embedPreview.html : null;
+
+    setMediaURL(value);
```
Comment (pablocubico) [inline assets/src/blocks/Media/MediaEditor.js:27]: As discussed by chat, this is already set in the onChange listener, it has to be there so the text is immediately updated in the input, otherwise there'll be a 300ms delay while typing.

File: assets/src/blocks/Media/MediaBlock.js [hunk 3]
```diff
@@ -17,6 +17,16 @@ const attributes = {
   },
 };
 
+// Old content didn't store the poster url and embed html in the props. For the frontend this is caught in the block's
+// backend render method, where we setup a div to be frontend rendered by the same MediaFrontend component as is used for
+// save.
+export const lacksAttributes = attributes => {
```
Comment (pablocubico) [inline assets/src/blocks/Media/MediaBlock.js:23]: Isn't this (and `patchLegacyAttributes` below) introducing more backwards compatibilty code than the current version? It looks pretty version-specific too, I think it's even harder to adapt or remove.

Comment (Inwerpsel) [reply assets/src/blocks/Media/MediaBlock.js:23]: It may be more lines, but they're much easier to remove than the effect:
https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/compare/media-change-listener...media-change-listener-remove-bc

Comment (Inwerpsel) [reply assets/src/blocks/Media/MediaBlock.js:23]: Also with this PR the BC code only affects content that actually needs it, while keeping our render code clean and not performing updates of the block's attributes, which is not a render concern.

Comment (Inwerpsel) [reply assets/src/blocks/Media/MediaBlock.js:23]: > It looks pretty version-specific too

Isn't BC code by nature version specific? :smile:

Comment (pablocubico) [reply assets/src/blocks/Media/MediaBlock.js:23]: Ha, I guess, but some can be more specific than others. So what's the next step here? Remove the WIP label?

Comment (Inwerpsel) [reply assets/src/blocks/Media/MediaBlock.js:23]: Sure, I'll just rebase it :+1:

File: assets/src/blocks/Media/MediaBlock.js [hunk 4]
```diff
@@ -36,9 +23,38 @@ export const registerMediaBlock = () => {
...
+    save: ({ attributes }) => {
+      if (lacksAttributes(attributes)) {
+        return null;
+      }
+
+      return <MediaFrontend { ...attributes } />;
+    },
     edit: MediaEditor,
-    mediaV1,
+    deprecated: [
```
Comment (Inwerpsel) [inline assets/src/blocks/Media/MediaBlock.js:56]: I don't think the previous code worked as `mediaV1` was only an array, not an object with the deprecated key.

File: assets/src/blocks/Media/MediaEditor.js (hunk 2 above)
Comment (Inwerpsel) [reply assets/src/blocks/Media/MediaEditor.js]: We ended up using the `defaultValue` value prop instead to resolve the issue

File: assets/src/blocks/Media/MediaFrontend.js [hunk 5]
```diff
@@ -19,7 +17,7 @@ const wrapEmbedHTML = embed_html => {
   return wrapperDiv.outerHTML;
 }
 
-export const MediaFrontend = ({ attributes }) => {
+export const MediaFrontend = ( attributes ) => {
```
Comment (Inwerpsel) [inline assets/src/blocks/Media/MediaFrontend.js:20]: Had to undo this change I suggested earlier on your PR, my bad! It needs to be matched in the frontend rendering code too, but then we would need to change the other frontend components, which is a bit too much for a PR that's already big.

**Review by pablocubico — APPROVED** [review]
Looks good, discussed by chat and tested locally.
